lib/snitch-core/src/lib.rs

```diff
@@ -0,0 +1,795 @@
+        self.insert_impl(sender, timestamp, data)
+    }
+
+    /// Iterates over the entire current contents of the store.
+    ///
+    /// This will produce `Record`s in the order they were received, with
+    /// ascending ENA values.
+    pub fn iter_contents(&mut self) -> impl Iterator<Item = Record<'_>> + '_ {
+        // Attempt to recover and insert a loss record if necessary.
+        self.recover_if_required(None);
+
+        let mut slices = self.storage.as_slices();
+        let mut next_ena = self.earliest_ena;
+        core::iter::from_fn(move || {
+            let len = u16::from_le_bytes(take_array(&mut slices)?) as usize;
+            let tid = u16::from_le_bytes(take_array(&mut slices)?);
+            let timestamp = u64::from_le_bytes(take_array(&mut slices)?);
+            let slices = take_slice(&mut slices, len)?;
+            let ena = next_ena;
+            next_ena += 1;
+            Some(Record {
+                ena,
+                tid,
+                timestamp,
+                slices,
+            })
+        })
+    }
+
+    /// Reads records starting from `first_ena`, inclusive.
+    pub fn read_from(
+        &mut self,
+        first_ena: u64,
+    ) -> impl Iterator<Item = Record<'_>> + '_ {
+        // We can use `skip_while` instead of `filter` here because the ENAs are
+        // always ascending.
+        self.iter_contents()
+            .skip_while(move |rec| rec.ena < first_ena)
+    }
+
+    /// Discards records up through and including `last_written_ena`.
+    ///
+    /// This is intended to be called when we've received confirmation that a
+    /// range of records has been committed to a database. The provided ENA is
+    /// _inclusive_ and indicates the last record written to the database.
+    ///
+    /// If the ENA is either lower than any of our stored records, or higher
+    /// than what we've vended out, this is a no-op.
+    pub fn flush_thru(&mut self, last_written_ena: u64) {
+        let Some(index) = last_written_ena.checked_sub(self.earliest_ena)
+        else {
+            // Cool, we're already aware that record has been written.
+            return;
+        };
+        if index >= self.stored_record_count as u64 {
+            // Uhhhh. We have not issued this ENA. It could not possibly have
+            // been written.
+            //
+            // TODO: is this an opportunity for the queue _itself_ to generate
+            // a defect report? For now, we'll just ignore it.
+            return;
+        }
+
+        for _ in 0..=index {
+            // Discard the lead record in the queue.
+            let mut slices = self.storage.as_slices();
+            let len = u16::from_le_bytes(take_array(&mut slices).unwrap_lite())
+                as usize;
+
+            let size = OVERHEAD + len;
+            // So it's weird, but, heapless::Deque only lets you pop individual
+            // bytes. Hopefully this is not too expensive.
```

Comment:
looks like it does [this](https://docs.rs/heapless/0.8.0/src/heapless/deque.rs.html#280-287)...

Reply:
Yeah, the question is what the code looks like when called in a loop, in practice, on our target. I haven't evaluated that yet. But it'll probably be fine.